src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingResult.cs

```diff
@@ -76,5 +84,54 @@ public ModelBindingResult(object model, string key, bool isModelSet, ModelValida
         /// A <see cref="ModelValidationNode"/> associated with the current <see cref="ModelBindingResult"/>.
         /// </summary>
         public ModelValidationNode ValidationNode { get; }
+
+        public override bool Equals(object obj)
+        {
+            var other = obj as ModelBindingResult?;
+            if (other == null)
+            {
+                return false;
+            }
+            else
+            {
+                return Equals(other.Value);
+            }
+        }
+
+        public override int GetHashCode()
+        {
+            if (Key == null)
+            {
+                return 0;
+            }
+            else if (IsModelSet)
+            {
+                Debug.Assert(Key != null);
+                return Key.GetHashCode() * Model?.GetHashCode() ?? 17;
+            }
+            else
+            {
+                Debug.Assert(Key != null);
+                return Key.GetHashCode();
+            }
+        }
+
+        public bool Equals(ModelBindingResult other)
+        {
+            return
+                string.Equals(Key, other.Key, StringComparison.Ordinal) &&
```

Comment:
Not sure about this. The comparable `ModelStateDictionary` keys are compared using `OrdinalIgnoreCase`. On the other hand, we don't have cases in existing code where `ModelBindingResult.Key` is ever compared (tests and that's it).